tests/components/smartthings/test_config_flow.py

```diff
@@ -8,434 +8,720 @@
+    refresh_token = str(uuid4())
+    smartthings_mock.apps.return_value = []
+    smartthings_mock.create_app.return_value = (app, app_oauth_client)
+    smartthings_mock.locations.return_value = [location]
+    request = Mock()
+    request.installed_app_id = installed_app_id
+    request.auth_token = token
+    request.location_id = location.location_id
+    request.refresh_token = refresh_token
 
-    result = await flow.async_step_pat()
+    with patch.object(
+        hass.components.cloud, "async_active_subscription", return_value=True
+    ), patch.object(
+        hass.components.cloud,
+        "async_create_cloudhook",
+        return_value=mock_coro("http://cloud.test"),
+    ) as mock_create_cloudhook:
 
-    assert flow.access_token == token
-    assert result["type"] == data_entry_flow.RESULT_TYPE_FORM
-    assert result["step_id"] == "pat"
-    assert result["errors"] == {}
-    assert result["data_schema"]({}) == {CONF_ACCESS_TOKEN: token}
-    assert "token_url" in result["description_placeholders"]
+        await smartapp.setup_smartapp_endpoint(hass)
+
+        # Webhook confirmation shown
+        result = await hass.config_entries.flow.async_init(
+            DOMAIN, context={"source": "user"}
+        )
+        assert result["type"] == data_entry_flow.RESULT_TYPE_FORM
+        assert result["step_id"] == "user"
+        assert result["description_placeholders"][
+            "webhook_url"
+        ] == smartapp.get_webhook_url(hass)
+        assert mock_create_cloudhook.call_count == 1
+
+        # Advance to PAT screen
+        result = await hass.config_entries.flow.async_configure(result["flow_id"], {})
+        assert result["type"] == data_entry_flow.RESULT_TYPE_FORM
+        assert result["step_id"] == "pat"
+        assert "token_url" in result["description_placeholders"]
+        assert "component_url" in result["description_placeholders"]
+
+        # Enter token and advance to location screen
+        result = await hass.config_entries.flow.async_configure(
+            result["flow_id"], {CONF_ACCESS_TOKEN: token}
+        )
+        assert result["type"] == data_entry_flow.RESULT_TYPE_FORM
+        assert result["step_id"] == "select_location"
+
+        # Select location and advance to external auth
+        result = await hass.config_entries.flow.async_configure(
+            result["flow_id"], {CONF_LOCATION_ID: location.location_id}
+        )
+        assert result["type"] == data_entry_flow.RESULT_TYPE_EXTERNAL_STEP
+        assert result["step_id"] == "authorize"
+        assert result["url"] == format_install_url(app.app_id, location.location_id)
+
+        # Complete external auth and advance to install
+        await smartapp.smartapp_install(hass, request, None, app)
+
+        # Finish
+        result = await hass.config_entries.flow.async_configure(result["flow_id"])
+        assert result["type"] == data_entry_flow.RESULT_TYPE_CREATE_ENTRY
+        assert result["data"]["app_id"] == app.app_id
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/foreign-sub/home-assistant/pullRequest?prid=5444210)